### ISSUE TITLE
Image scaling bug

### ISSUE THREAD

**Issue by prabhuamol** [kean/Nuke#791]
We are sometimes noticing when images are scaled using .thumbnailKey they appear blurry. I wonder if its cause in` func makeThumbnail(data: Data, options: ImageRequest.ThumbnailOptions) -> PlatformImage?` scale is not taken into account when returning back the UIImage. Which is causing the dimensions of the image to be reported incorrectly.

We do some post processing via `prepareThumbnail` if the returned thumbnail size does not match the view size exactly and the scaled down image returned using https://developer.apple.com/documentation/uikit/uiimage/3750845-preparethumbnailofsize is too blurry.

Heres a sample request and response to show the issue

▿ ImageRequest(resource: "some url (hidden for privacy)", priority: normal, processors: [], options: Options(rawValue: 0), userInfo: [Nuke.ImageRequest.UserInfoKey(rawValue: "github.com/kean/nuke/thumbnail"): Nuke.ImageRequest.ThumbnailOptions(targetSize: Nuke.ImageRequest.ThumbnailOptions.TargetSize.fixed(474.0), createThumbnailFromImageIfAbsent: true, createThumbnailFromImageAlways: true, createThumbnailWithTransform: true, shouldCacheImmediately: true), Nuke.ImageRequest.UserInfoKey(rawValue: "imageSize"): (110.0, 158.0)])
  ▿ ref : <Container: 0x600001819c40>
  
 
(lldb) po response.image
<UIImage:0x6000030307e0 anonymous {338, 474} renderingMode=automatic(original)>

(lldb) po response.image.size
▿ (338.0, 474.0)
  - width : 338.0
  - height : 474.0

As you can see image size is reported in pixel size and not points. The fix might be that when we return PlatformImage in Nuke.makeThumbnail use scale initializer for UIImage. https://github.com/kean/Nuke/blob/084076db4f9b9ee4db51177ef5a4879689c1aa09/Sources/Nuke/Internal/Graphics.swift#L369

`return PlatformImage(cgImage: image, scale: scale, orientaion: orientation)` instead of return `PlatformImage(cgImage: image)` I had to do that in an app i was working on as well

**Comment by prabhuamol** [comment]
You can also verify this by just chaning any of the `ImageProcessorsResizeTests` to use points and all of them fail

<img width="1728" alt="Screenshot 2024-06-14 at 2 50 28 PM" src="https://github.com/kean/Nuke/assets/3059888/8fcf4527-1442-47b1-87b7-e04c95d29a34">

I also ran the `testThatImageIsResized` and verified that image size and sizeInPixels is the same. Where one should be in points and smaller

<img width="1723" alt="Screenshot 2024-06-14 at 2 53 56 PM" src="https://github.com/kean/Nuke/assets/3059888/7e07ea0e-96aa-4fc9-b11b-06d914177412">

**Comment by prabhuamol** [comment]
@kean any luck investigating this? Would be amazing to get this fixed

**Comment by prabhuamol** [comment]
Submitted a PR here https://github.com/kean/Nuke/pull/793

**Comment by kean** [comment]
Hey, @prabhuamol! 

I've looked into it, and I might need some clarification as I'm not sure I follow this entirely.

By default, all images fetched by Nuke are returned with a scale of `1`, so it's expected that `image.size` will return size in pixels. The framework doesn't know and can't know what is the logical size of the images it fetches. 

For example, let's say you download an icon for a button from a server. The server returns an image with a resolution of 88×88 pixels but a logical size of 44×44 pixels. In the iOS terminology, it will be an image with a scale of `2`. If you fetch it on a device with a `UIScreen` scale of `3`, the image should still have a scale of `2`.

If you need to change the scale of the feched image, you can use [ImageRequest.UserInfoKey.scaleKey](https://kean-docs.github.io/nuke/documentation/nuke/imagerequest/userinfokey/scalekey). But the way I typically approach it is by never relying on the logical scale of the images downloaded from the internet and instead restricting the point size of the views in which they are displayed.

> You can also verify this by just chaning any of the ImageProcessorsResizeTests to use points and all of them fail

It works as expected. When you change the `unit` to `.points`, it simply means "resize the image to a width `400 × UIScreen.main.scale` pixels". On a 3x device, it will be 1200 pixels, but `ImageProcessors.Resize` never upscales images by deafult, so you end up with an image that is 640 pixels wide. The `unit` has nothing to do with the `scale` of the output images or their logical size.

**Comment by prabhuamol** [comment]
Hey @kean So the issue is that the returned UIImage from resize process does not take device scale into account. So lets say i wanted to resize an image to CGSize of 44 on a device with scale of 2. The returned image has pixel size of 88*88 which makes sense but associated UIImage.size is also 88*88 (which is not correct) The size of the UIImage should take the device scale into account and should not be a 1:1 mapping to the CGImage pixel size. The PR i submitted here fixes that https://github.com/kean/Nuke/pull/793 and also adds related tests. Also the scaleKey doc says that by default it matches the scale of the display which seems to be not true based on your comment above . 
> By default, all images fetched by Nuke are returned with a scale of 1

Currently you can also see the tests you have in the app will fail as i have shown in the screenshot above. Hope this helps

**Comment by kean** [comment]
> Also the scaleKey doc says that by default it matches the scale of the display which seems to be not true based on your comment above .

Good catch. I updated it to the following:

```
 /// The image scale to be used. By default, the scale is `1`.
```

**Comment by kean** [comment]
I'd also suggest checking the Apple documentation about `UIImage` [scale](https://developer.apple.com/documentation/uikit/uiimage/1624110-scale). I also used to think it should generally always match the scale of the screen, but that's not the case. The default scale for unknown images should be `1`.

**Comment by prabhuamol** [comment]
> e the tests you have in the app will fail as i have show

I read the doc but I think that is very specific to images inside a bundle. So can you point us to how do we make sure that the resized image we get back is returned with the right point size information? i.e the image is getting resized correctly but its just not returning the  point size information accurately (i feel like this is a bug).  It should follow the following logic as far as i can tell UIImage.size (points) = CGImage.size (pixels) * UIScreen.main.scale. I feel like the scale information is just lost when we resize images. So from consumer side the image returned is "larger" than it should be because the size info is not correct and is always matching 1:1 to pixel size irrespective of the device. 

On a related note i dont see the scale key being uses anywhere in the thumbnail creation in `func makeThumbnail(data: Data, options: ImageRequest.ThumbnailOptions) -> PlatformImage?` Is that a way for us to specify the scale and get the right size back? Thanks!

**Comment by kean** [comment]
> On a related note i dont see the scale key being uses anywhere in the thumbnail creation in func makeThumbnail(data: Data, options: ImageRequest.ThumbnailOptions) -> PlatformImage? Is that a way for us to specify the scale and get the right size back? Thanks!

This is a bug, and PRs with a fix would be more than welcome.

In the future major version, I have it on my list to see if the scale support could be moved to the later stages of the pipeline. For example ,`AsyncImage` lets you [change it](https://developer.apple.com/documentation/swiftui/asyncimage/init(url:scale:)) during the _display_ phase – very later.

**Comment by prabhuamol** [comment]
> > On a related note i dont see the scale key being uses anywhere in the thumbnail creation in func makeThumbnail(data: Data, options: ImageRequest.ThumbnailOptions) -> PlatformImage? Is that a way for us to specify the scale and get the right size back? Thanks!
> 
> This is a bug, and PRs with a fix would be more than welcome.
> 
> In the future major version, I have it on my list to see if the scale support could be moved to the later stages of the pipeline. Applying it early has implications for caching and request coalescing.
> 
> For example ,`AsyncImage` lets you [change it](https://developer.apple.com/documentation/swiftui/asyncimage/init(url:scale:)) during the _display_ phase – very later.

PR is here https://github.com/kean/Nuke/pull/793

**Comment by kean** [comment]
Fixed in [v12.8.0](https://github.com/kean/Nuke/releases/tag/12.8.0). Thanks @prabhuamol for sticking with in and getting a PR out.
